crates/order-validation/src/banned.rs

```diff
@@ -1,22 +1,122 @@
 use {
-    cached::{Cached, TimedCache},
     contracts::ChainalysisOracle,
     ethcontract::{errors::MethodError, futures::future::join_all, H160},
-    std::{collections::HashSet, sync::Mutex},
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+        time::{Duration, Instant},
+    },
+    tokio::sync::RwLock,
 };
 
 /// A list of banned users and an optional registry that can be checked onchain.
 pub struct Users {
     list: HashSet<H160>,
-    onchain: Option<Onchain>,
+    onchain: Option<Arc<Onchain>>,
+}
+
+#[derive(Clone)]
+struct UserMetadata {
+    is_banned: bool,
+    last_updated: Instant,
+}
+
+impl UserMetadata {
```

Comment:
`with_banned()` is not used at all and `with_last_update()` only once. I think it's reasonable to drop the `impl` block altogether and do a "manual" update a la:
```rust
let updated = UserMetadata {
    last_update: new_value,
    ..original_metadata
}
```